[Help Request] Evaluating the trained RL model errors

### High Level Description

I am having difficulty evaluating the trained RL model.

### Version

Master branch [291417](https://github.com/huawei-noah/SMARTS/commit/291417240070af7cfd0773815a5ceec7005caf4b)

### Operating System

ubuntu 20.04

### Problems

I followed [driving_smarts_2023_3](https://smarts.readthedocs.io/en/latest/benchmarks/driving_smarts_2023_3.html) and trained a rl model. 

When I want to evaluate the model, output errors as followed:
```
<-- Starting `Driving SMARTS 2023.3` benchmark -->

This is the Driving SMARTS 2023.3 benchmark.

For context see:
    - https://smarts-project.github.io/competition/2023_driving_smarts/
    - https://codalab.lisn.upsaclay.fr/competitions/


Evaluating platoon-scenarios/sumo/platoon/merge_exit_sumo_t_agents_1...
Using configuration from: /home/sutpc/lym/SMARTS/smarts/engine.ini
:display:egldisplay(error): Failed to find or initialize a suitable EGL display connection.
Attempt to register type x11GraphicsPipe more than once!
Attempt to register type x11GraphicsWindow more than once!
Xlib:  extension "XFree86-DGA" missing on display "localhost:10.0".
Xlib:  extension "XFree86-DGA" missing on display "localhost:10.0".
Traceback (most recent call last):
  File "/home/sutpc/lym/SMARTS/smarts/benchmark/entrypoints/benchmark_runner_v0.py", line 84, in _eval_worker_local
    obs, reward, terminated, truncated, info = env.step(action)
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/gymnasium/core.py", line 361, in step
    return self.env.step(action)
  File "/home/sutpc/lym/SMARTS/smarts/env/gymnasium/wrappers/metric/metrics.py", line 130, in step
    new_costs = cost_func(
  File "/home/sutpc/lym/SMARTS/smarts/env/gymnasium/wrappers/metric/costs.py", line 411, in func
    assert min_len >= min_waypoints_length, (
AssertionError: Expected waypoints length >= 28, but got waypoints length = 25.

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/home/sutpc/miniconda3/envs/rl/bin/scl", line 8, in <module>
    sys.exit(scl())
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/click/core.py", line 1128, in __call__
    return self.main(*args, **kwargs)
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/click/core.py", line 1053, in main
    rv = self.invoke(ctx)
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/click/core.py", line 1659, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/click/core.py", line 1395, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/home/sutpc/miniconda3/envs/rl/lib/python3.8/site-packages/click/core.py", line 754, in invoke
    return __callback(*args, **kwargs)
  File "/home/sutpc/lym/SMARTS/cli/benchmark.py", line 79, in run
    run_benchmark(
  File "/home/sutpc/lym/SMARTS/smarts/benchmark/__init__.py", line 109, in run_benchmark
    entrypoint(
  File "/home/sutpc/lym/SMARTS/smarts/benchmark/entrypoints/benchmark_runner_v0.py", line 221, in benchmark_from_configs
    benchmark(
  File "/home/sutpc/lym/SMARTS/smarts/benchmark/entrypoints/benchmark_runner_v0.py", line 174, in benchmark
    for name, records in iterator(
  File "/home/sutpc/lym/SMARTS/smarts/benchmark/entrypoints/benchmark_runner_v0.py", line 126, in _serial_task_iterator
    name, records = _eval_worker_local(
  File "/home/sutpc/lym/SMARTS/smarts/benchmark/entrypoints/benchmark_runner_v0.py", line 89, in _eval_worker_local
    records = env.records()
  File "/home/sutpc/lym/SMARTS/smarts/env/gymnasium/wrappers/metric/metrics.py", line 300, in records
    costs=op_dataclass(
  File "/home/sutpc/lym/SMARTS/smarts/env/gymnasium/wrappers/metric/utils.py", line 70, in op_dataclass
    new[field.name] = op(getattr(first, field.name), second)
  File "/home/sutpc/lym/SMARTS/smarts/env/gymnasium/wrappers/metric/utils.py", line 86, in divide
    return float(value / divider)
ZeroDivisionError: division by zero
WARNING:SumoTrafficSimulation:TraCI has provided a warning Vehicle 'Agent_0' is not known

```

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -44,6 +44,7 @@ Copy and pasting the git commit messages is __NOT__ enough.
 - Fixed an issue where a single waypoint would appear in off-route lanes for missions with a goal.
 - Fixed an underlying issue with the sensor manager where the sensors were not removed immediately.
 - Fixed an issue where warnings could be generated when an agent takes over an existing vehicle if the vehicle previously had sensors on it.
+- Fixed vehicle gap metric to consume ragged waypoints matrix.
 ### Removed
 - Removed the following dependencies from smarts: `pandas`, `rich`, `twisted`, `sh`.
 - Moved `baselines/marl_benchmark` from this repository to `smarts-project/smarts-project.rl` repository.

--- a/smarts/env/gymnasium/wrappers/metric/costs.py
+++ b/smarts/env/gymnasium/wrappers/metric/costs.py
@@ -19,7 +19,6 @@
 # THE SOFTWARE.
 from __future__ import annotations
 
-import warnings
 from collections import deque
 from dataclasses import dataclass
 from typing import Callable, Dict, List, NewType, Optional, Tuple
@@ -461,6 +460,9 @@ def _vehicle_gap(
     # vehicle separation distance while driving.
     min_waypoints_length = int(np.ceil(max_column_length / waypoint_spacing))
 
+    def convert_2d_to_3d(array: np.ndarray) -> np.ndarray:
+        return np.pad(array, (0, 1), mode="constant", constant_values=0)
+
     def func(
         road_map: RoadMap, vehicle_index: VehicleIndex, done: Done, obs: Observation
     ) -> Costs:
@@ -479,25 +481,25 @@ def _vehicle_gap(
             max_column_length,
         )
 
-        # Truncate all paths to be of the same length.
-        min_len = min(map(len, obs.waypoint_paths))
-        if min_len >= min_waypoints_length:
-            warnings.warn(
-                f"Expected waypoints length >= {min_waypoints_length}, but got "
-                + f"waypoints length = {min_len}."
-            )
-        trunc_waypoints = list(map(lambda x: x[:min_len], obs.waypoint_paths))
-        waypoints = [list(map(lambda x: x.pos, path)) for path in trunc_waypoints]
-        waypoints = np.array(waypoints, dtype=np.float64)
-        waypoints = np.pad(
-            waypoints, ((0, 0), (0, 0), (0, 1)), mode="constant", constant_values=0
-        )
+        # fmt: off
+        waypoint_paths = obs.waypoint_paths
+        max_len = max(map(len, waypoint_paths))
+        mask = np.array([
+            [[False, False, False]] * len(path) + [[True, True, True]] * (max_len - len(path))
+            for path in waypoint_paths
+        ])
+        waypoints = np.array([
+            list(map(lambda x: convert_2d_to_3d(x.pos), path)) + [np.full((3,), np.nan)] * (max_len - len(path))
+            for path in waypoint_paths
+        ])
+        waypoints_masked = np.ma.MaskedArray(data=waypoints, mask=mask)
+        # fmt: on
 
         # Find the nearest waypoint index to the actor of interest, if any.
-        lane_width = obs.waypoint_paths[0][0].lane_width
+        lane_width = waypoint_paths[0][0].lane_width
         aoi_pos = vehicle_index.vehicle_position(aoi)
         aoi_wp_ind, aoi_ind = nearest_waypoint(
-            matrix=waypoints,
+            matrix=waypoints_masked,
             points=np.array([aoi_pos]),
             radius=lane_width,
         )
@@ -511,12 +513,12 @@ def _vehicle_gap(
         else:
             # Find the nearest waypoint index to the ego.
             ego_pos = obs.ego_vehicle_state.position
-            dist = np.linalg.norm(waypoints[:, 0, :] - ego_pos, axis=-1)
+            dist = np.linalg.norm(waypoints_masked[:, 0, :] - ego_pos, axis=-1)
             ego_wp_inds = np.where(dist == dist.min())[0]
 
             if aoi_wp_ind[0] in ego_wp_inds:
                 # Ego is in the same lane as the actor of interest.
-                j_gap = (aoi_wp_ind[1] * waypoint_spacing - vehicle_length) / (
+                j_gap = max(aoi_wp_ind[1] * waypoint_spacing - vehicle_length, 0) / (
                     column_length - vehicle_length
                 )
             else:

--- a/smarts/env/gymnasium/wrappers/metric/utils.py
+++ b/smarts/env/gymnasium/wrappers/metric/utils.py
@@ -100,7 +100,7 @@ def multiply(value: Union[int, float], multiplier: Union[int, float]) -> float:
 
 
 def nearest_waypoint(
-    matrix: np.ndarray, points: np.ndarray, radius: float = 1
+    matrix: np.ma.MaskedArray, points: np.ndarray, radius: float = 1
 ) -> Tuple[Tuple[int, int], Optional[int]]:
     """
     Returns
@@ -110,14 +110,14 @@ def nearest_waypoint(
     Nearby is defined as a point within `radius` of a waypoint.
 
     Args:
-        matrix (np.ndarray): Waypoints matrix.
+        matrix (np.ma.MaskedArray): Waypoints matrix.
         points (np.ndarray): Points matrix.
         radius (float, optional): Nearby radius. Defaults to 2.
 
     Returns:
         Tuple[Tuple[int, int], Optional[int]] : `matrix` index of shape (a,b) and scalar `point` index.
     """
-    cur_point_index = ((np.intp(1e10), np.intp(1e10)), None)
+    cur_point_index = ((np.int(1e10), np.int(1e10)), None)
 
     if points.shape == (0,):
         return cur_point_index
@@ -130,10 +130,11 @@ def nearest_waypoint(
     points_expanded = np.expand_dims(points, (1, 2))
     diff = matrix - points_expanded
     dist = np.linalg.norm(diff, axis=-1)
+    dist_masked = np.ma.MaskedArray(dist, diff.mask[..., 0])
     for ii in range(points.shape[0]):
-        index = np.argmin(dist[ii])
-        index_unravel = np.unravel_index(index, dist[ii].shape)
-        min_dist = dist[ii][index_unravel]
+        index = np.argmin(dist_masked[ii])
+        index_unravel = np.unravel_index(index, dist_masked[ii].shape)
+        min_dist = dist_masked[ii][index_unravel]
         if min_dist <= radius and index_unravel[1] < cur_point_index[0][1]:
             cur_point_index = (index_unravel, ii)
 

